pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -0,0 +1,403 @@
+	nodeLister := apiClients.NodeInformer.Lister()
+
+	rc := recorder.WithComponentSuffix("vmware-" + strings.ToLower(name))
+
+	c := &VSphereController{
+		name:            name,
+		targetNamespace: targetNamespace,
+		kubeClient:      apiClients.KubeClient,
+		operatorClient:  apiClients.OperatorClient,
+		configMapLister: configMapInformer.Lister(),
+		secretLister:    apiClients.SecretInformer.Lister(),
+		csiNodeLister:   csiNodeLister,
+		scLister:        scInformer.Lister(),
+		csiDriverLister: csiDriverLister,
+		nodeLister:      nodeLister,
+		apiClients:      apiClients,
+		eventRecorder:   rc,
+		vSphereChecker:  newVSphereEnvironmentChecker(),
+		infraLister:     infraInformer.Lister(),
+	}
+	c.controllers = []conditionalController{}
+	c.createCSIDriver()
+	c.createWebHookController()
+	c.storageClassController = c.createStorageClassController()
+
+	return factory.New().WithInformers(
+		apiClients.OperatorClient.Informer(),
+		configMapInformer.Informer(),
+		apiClients.SecretInformer.Informer(),
+		infraInformer.Informer(),
+		scInformer.Informer(),
+	).WithSync(c.sync).
+		ResyncEvery(resyncDuration).
+		WithSyncDegradedOnError(apiClients.OperatorClient).ToController(c.name, rc)
+}
+
+func (c *VSphereController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	klog.V(4).Infof("%s: sync started", c.name)
+	defer klog.V(4).Infof("%s: sync complete", c.name)
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != operatorapi.Managed {
+		klog.V(4).Infof("%s: ManagementState is %s, skipping", c.name, opSpec.ManagementState)
+		return nil
+	}
+
+	infra, err := c.infraLister.Get(infraGlobalName)
+	if err != nil {
+		return err
+	}
+
+	if infra.Status.PlatformStatus == nil {
+		klog.V(4).Infof("Unknown platform: infrastructure status.platformStatus is nil")
+		return nil
+	}
+
+	if infra.Status.PlatformStatus.Type != ocpv1.VSpherePlatformType {
+		klog.V(4).Infof("Unsupported platform: infrastructure status.platformStatus.type is %s", infra.Status.PlatformStatus.Type)
+		return nil
+	}
+
+	driverCheckFlag, err := c.driverAlreadyStarted(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if driver was previously started, then start it even if checks are failing
+	if driverCheckFlag && !c.operandControllerStarted {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+
+	connectionResult := c.loginToVCenter(ctx, infra)
+	defer func() {
+		klog.V(4).Infof("%s: vcenter-csi logging out from vcenter", c.name)
+		err := c.vSphereConnection.Logout(ctx)
+		if err != nil {
+			klog.Errorf("%s: error closing connection to vCenter API: %v", c.name, err)
+		}
+	}()
+
+	// if there was an OCP error we should degrade the cluster or if we previously created CSIDriver
+	// but we can't connect to vcenter now, we should also degrade the cluster
+	err, degradedOrBlockedFromUpgrades := c.blockUpgradeOrDegradeCluster(ctx, connectionResult, infra)
```

Comment:
`blockUpgradeOrDegradeCluster` is called a lot in this function and it will always update the ClusterCSIDriver conditions.

* This is quite aggressive.
* When e.g.  `loginToVCenter` checks succeeds, the conditions get cleared, but when a subsequent check fails, they get filed again - they will be flipping too much IMO.

Should there be a single function that does `sync()` and returns the final conditions that are applied only once?

Reply:
hmm, if ` loginToVCenter` check succeeds the conditions are not touched at all. Conditions are only updated if either `blockUpgrade` is true.

Reply:
fixed.